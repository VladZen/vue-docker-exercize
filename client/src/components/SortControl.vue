<template>
  <div class="sort-control">
    <span>Sorted by {{ sortBy.column }}:</span>

    <select v-model="sortBy" class="sort-control__select">
      <optgroup v-for="column in sortings" :key="column" :label="column">
        <option
          v-for="direction in dirs"
          :value="{ column, direction }"
          :key="direction"
        >
          {{ direction }}
        </option>
      </optgroup>
    </select>
  </div>
</template>

<script>
import { mapMutations, mapState } from "vuex";

export default {
  data() {
    return {
      sortings: ["name", "quantity"],
      dirs: ["asc", "desc"],
    };
  },
  computed: {
    ...mapState(["orderBy"]),
    sortBy: {
      get() {
        return this.orderBy;
      },
      set(value) {
        this.setOrder(value);
      },
    },
  },
  methods: {
    ...mapMutations(["setOrder"]),
  },
};
</script>

<style lang="scss" scoped>
.sort-control {
  display: flex;
  justify-content: flex-end;
  margin-bottom: $base-offset;

  &__select {
    margin-left: $base-offset;
  }
}
</style>
