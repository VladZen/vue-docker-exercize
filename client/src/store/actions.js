import axios from "axios";

const _HELPERS = {
  makeFormData(object) {
    const formData = new FormData();
    Object.keys(object).forEach((key) => {
      formData.append(key, object[key]);
    });
    return formData;
  },
  cleanData(obj) {
    return Object.keys(obj).reduce((result, key) => {
      if (typeof obj[key] !== "undefined" && obj[key] !== null) {
        result[key] = obj[key];
      }
      return result;
    }, {});
  },
  notInList(list, item) {
    return !list.some(({ id }) => item.id === id);
  },
  getFromList(list, itemId) {
    return list.find(({ id }) => itemId === id);
  },
};

const actions = {
  async GET_MERCHANT({ commit, state }) {
    if (state.merchantName) return state.merchantName;
    const { data: merchantName } = await axios.get("/api/merchant-name/");
    commit("setMerchantName", merchantName);
  },
  async GET_INVENTORY({ commit }) {
    const { data: inventory } = await axios.get("/api/inventory/");
    commit("initializeList", inventory);
  },
  async GET_ITEM({ commit, state }, id) {
    const item = _HELPERS.getFromList.call(null, state.list, id);
    if (item) return item;
    const { data: newItem } = await axios.get(`/api/inventory/${id}/`);
    commit("addItem", newItem);
    return newItem;
  },
  async REMOVE({ commit, state }, item) {
    if (_HELPERS.notInList.call(null, state.list, item)) return;
    await axios.delete(`/api/inventory/${item.id}/`);
    commit("removeItem", item.id);
  },
  async ADD({ commit }, name) {
    const formData = _HELPERS.makeFormData({ name });
    const { data: item } = await axios.post("/api/inventory/", formData);
    commit("addItem", item);
    return item.id;
  },
  async UPDATE({ commit, state }, { id, name, quantity }) {
    if (_HELPERS.notInList.call(null, state.list, { id })) return;
    const cleanData = _HELPERS.cleanData({ name, quantity });
    const formData = _HELPERS.makeFormData(cleanData);
    await axios.patch(`/api/inventory/${id}/`, formData);

    const newItem = { id, ...cleanData };
    commit("updateItem", newItem);
    return newItem;
  },
};

export { actions };
