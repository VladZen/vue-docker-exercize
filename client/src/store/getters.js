import _sortBy from "lodash.orderby";

const getters = {
  sortedList({ list, orderBy: { direction, column } }) {
    return _sortBy(list, [column], [direction]);
  },
};

export { getters };
