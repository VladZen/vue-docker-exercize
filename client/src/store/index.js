import Vue from "vue";
import Vuex from "vuex";
import axios from "axios";

import defaultItem from "@/config/default-item";

Vue.use(Vuex);

const _HELPERS = {
  makeFormData(object) {
    const formData = new FormData();
    Object.keys(object).forEach((key) => {
      formData.append(key, object[key]);
    });
    return formData;
  },
  cleanData(obj) {
    return Object.keys(obj).reduce((result, key) => {
      if (typeof obj[key] !== "undefined" && obj[key] !== null) {
        result[key] = obj[key];
      }
      return result;
    }, {});
  },
  notInList(list, item) {
    return !list.some(({ id }) => item.id === id);
  },
  getFromList(list, itemId) {
    return list.find(({ id }) => itemId === id);
  },
};

export default new Vuex.Store({
  state: () => ({
    merchantName: "",
    list: [],
  }),
  mutations: {
    initializeList(state, list) {
      state.list = [].concat(list);
    },
    updateItem(state, item) {
      state.list = state.list.map((listItem) =>
        item.id === listItem.id ? item : listItem
      );
    },
    addItem(state, item = defaultItem()) {
      state.list.unshift(item);
    },
    removeItem(state, itemId) {
      state.list = state.list.filter(({ id }) => id !== itemId);
    },
    setMerchantName(state, merchantName) {
      state.merchantName = merchantName;
    },
  },
  actions: {
    async GET_MERCHANT({ commit }) {
      const { data: merchantName } = await axios.get("/api/merchant-name/");
      commit("setMerchantName", merchantName);
    },
    async GET_INVENTORY({ commit }) {
      const { data: inventory } = await axios.get("/api/inventory/");
      commit("initializeList", inventory);
    },
    async GET_ITEM({ commit, state }, id) {
      const item = _HELPERS.getFromList.call(null, state.list, id);
      if (item) return item;
      const { data: newItem } = await axios.get(`/api/inventory/${id}/`);
      commit("addItem", newItem);
      return newItem;
    },
    async REMOVE({ commit, state }, item) {
      if (_HELPERS.notInList.call(null, state.list, item)) return;
      await axios.delete(`/api/inventory/${item.id}/`);
      commit("removeItem", item.id);
    },
    async ADD({ commit }, name) {
      const formData = _HELPERS.makeFormData({ name });
      const item = await axios.post("/api/inventory/", formData);
      commit("addItem", item);
      return item.id;
    },
    async UPDATE({ commit, state }, { id, name, quantity }) {
      if (_HELPERS.notInList.call(null, state.list, { id })) return;
      const cleanData = _HELPERS.cleanData({ name, quantity });
      const formData = _HELPERS.makeFormData(cleanData);
      await axios.patch(`/api/inventory/${id}/`, formData);

      const newItem = { id, ...cleanData };
      commit("updateItem", newItem);
      return newItem;
    },
  },
});
