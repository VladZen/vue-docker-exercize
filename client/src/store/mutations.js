const mutations = {
  initializeList(state, list) {
    state.list = [].concat(list);
  },
  updateItem(state, item) {
    state.list = state.list.map((listItem) =>
      item.id === listItem.id ? item : listItem
    );
  },
  addItem(state, item) {
    state.list.unshift(item);
  },
  removeItem(state, itemId) {
    state.list = state.list.filter(({ id }) => id !== itemId);
  },
  setMerchantName(state, merchantName) {
    state.merchantName = merchantName;
  },
  setOrder(state, orderObject) {
    state.orderBy = {
      ...state.orderBy,
      ...orderObject,
    };
  },
};

export { mutations };
