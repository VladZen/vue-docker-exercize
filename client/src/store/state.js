const state = () => ({
  merchantName: "",
  list: [],
  orderBy: {
    column: "name",
    direction: "asc",
  },
});

export { state };
