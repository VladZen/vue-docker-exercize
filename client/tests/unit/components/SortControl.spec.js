import { createLocalVue, mount } from "@vue/test-utils";
import SortControl from "@/components/SortControl";
import Vuex from "vuex";
import { mutations } from "@/store/mutations";

const localVue = createLocalVue();
localVue.use(Vuex);

const store = new Vuex.Store({
  state: () => ({ orderBy: { column: "name", direction: "asc" } }),
  mutations: {
    setOrder: mutations.setOrder,
  },
});

const mountSortControl = () => mount(SortControl, { localVue, store });

describe("SortControl", () => {
  let wrapper;

  const select = () => wrapper.find(".sort-control__select");

  beforeAll(() => {
    wrapper = mountSortControl();
  });

  it("checks basic rendering", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("checks manipulations with the select", async () => {
    const options = select().findAll("option");
    await options.at(1).setSelected();
    expect(wrapper).toMatchSnapshot();
  });
});
