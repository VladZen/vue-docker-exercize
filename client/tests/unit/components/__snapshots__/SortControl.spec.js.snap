// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortControl checks basic rendering 1`] = `
<div class="sort-control"><span>Sorted by name:</span> <select class="sort-control__select">
    <optgroup label="name">
      <option value="[object Object]">
        asc
      </option>
      <option value="[object Object]">
        desc
      </option>
    </optgroup>
    <optgroup label="quantity">
      <option value="[object Object]">
        asc
      </option>
      <option value="[object Object]">
        desc
      </option>
    </optgroup>
  </select></div>
`;

exports[`SortControl checks manipulations with the select 1`] = `
<div class="sort-control"><span>Sorted by name:</span> <select class="sort-control__select">
    <optgroup label="name">
      <option value="[object Object]">
        asc
      </option>
      <option value="[object Object]">
        desc
      </option>
    </optgroup>
    <optgroup label="quantity">
      <option value="[object Object]">
        asc
      </option>
      <option value="[object Object]">
        desc
      </option>
    </optgroup>
  </select></div>
`;
